src/array.cr

```diff
@@ -2029,6 +2029,36 @@ class Array(T)
     @capacity - @offset_to_buffer
   end
 
+  private def rewind
+    root_buffer.copy_from(@buffer, @size)
+    shift_buffer_by -@offset_to_buffer
+  end
+
+  # Enusures that the internal buffer has at least `capacity` elements. If
+  # `rewind` is truthy, then the internal buffer is rewind first: the
+  # elements are copied to the front of the array.
+  def ensure_capacity(capacity : Int32, rewind = true) : self
+    rewind() if rewind
+
+    if capacity - @offset_to_buffer >= @size
+      resize_to_capacity capacity
+    else
+      err = "Insufficient capacity: #{capacity} cannot hold #{@size} elements"
+      err += " with an offset of #{@offset_to_buffer}" if @offset_to_buffer > 0
```

Comment:
Similar to the discussion in https://github.com/crystal-lang/crystal/pull/11485/files#r756340774, this offset part of the error message is an internal implementation detail that users will likely not understand.